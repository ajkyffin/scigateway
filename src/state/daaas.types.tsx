import { Step } from 'react-joyride';

export const AuthSuccessType = 'daaas:auth_success';
export const AuthFailureType = 'daaas:auth_failure';
export const LoadingAuthType = 'daaas:loading_auth';
export const LoadAuthProviderType = 'daaas:loading_auth_provider';
export const TokenExpiredType = 'daaas:expired_token';
export const ConfigureStringsType = 'daaas:configure_strings';
export const LoginType = 'daaas:login';
export const NotificationType = 'daaas:api:notification';
export const RegisterRouteType = 'daaas:api:register_route';
export const RequestPluginRerenderType = 'daaas:api:plugin_rerender';
export const SignOutType = 'daaas:signout';
export const ToggleDrawerType = 'daaas:toggledrawer';
export const DismissNotificationType = 'daaas:dismissnotification';
export const ConfigureFeatureSwitchesType = 'daaas:feature_switches';
export const AddNotificationsType = 'daaas:add_notification';
export const SiteLoadingType = 'daaas:site_loading';
export const ConfigureAnalyticsType = 'daaas:configure_analytics';
export const InitialiseAnalyticsType = 'daaas:initialise_analytics';
export const ToggleHelpType = 'daaas:toggle_help';
export const AddHelpTourStepsType = 'daaas:add_help_tour_steps';

export interface NotificationPayload {
  message: string;
  severity: string;
}

export interface AddNotificationsPayload {
  notifications: NotificationPayload[];
}

export interface AppStrings {
  [id: string]: string;
}

export interface ConfigureStringsPayload {
  res: ApplicationStrings;
}

export interface ApplicationStrings {
  [section: string]: AppStrings;
}

export interface FeatureSwitchesPayload {
  switches: FeatureSwitches;
}

export interface FeatureSwitches {
  showContactButton: boolean;
}

export interface RegisterRoutePayload {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
}

export interface PluginConfig {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
}

export interface GroupedPlugins {
  [section: string]: PluginConfig[];
}

export interface LoginPayload {
  username: string;
  password: string;
}

export interface AccessPayload {
  authorisation: boolean;
  accessToken: string;
  error: null;
}

export interface AuthorisedPayload {
  token: string;
}

export interface DismissNotificationPayload {
  index: number;
}

export interface AuthProviderPayload {
  authProvider: string;
}

export interface SiteLoadingPayload {
  loading: boolean;
}

export interface ConfigureAnalyticsPayload {
  id: string;
}

export interface AddHelpTourStepsPayload {
  steps: Step[];
}
