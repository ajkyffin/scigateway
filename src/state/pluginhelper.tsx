import {
  GroupedPlugins,
  PluginConfig,
  RegisterRoutePayload,
} from './daaas.types';

// Custom comparer for plugin objects, sorts keys by order (asc), then by name (asc)
export const comparePlugins = (a: PluginConfig, b: PluginConfig): number =>
  a.order === b.order
    ? a.displayName.localeCompare(b.displayName)
    : a.order - b.order;

function buildMenuItemDictionary(plugins: PluginConfig[]): GroupedPlugins {
  const dict: GroupedPlugins = {};

  plugins.forEach(p => {
    if (!(p.section in dict)) {
      dict[p.section] = [];
    }
    dict[p.section].push(p);
  });

  return dict;
}

// Convert the list of plugins into a structured dataset for rendering
export function structureMenuData(plugins: PluginConfig[]): GroupedPlugins {
  const menuItems = buildMenuItemDictionary(plugins);

  Object.keys(menuItems).forEach(
    section => (menuItems[section] = menuItems[section].sort(comparePlugins))
  );
  return menuItems;
}

// Convert a plugin registration request body into a pluginConfig object
export function buildPluginConfig(payload: RegisterRoutePayload): PluginConfig {
  return { ...payload };
}
