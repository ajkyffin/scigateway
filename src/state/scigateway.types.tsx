import { Step } from 'react-joyride';

export const AuthSuccessType = 'scigateway:auth_success';
export const AuthFailureType = 'scigateway:auth_failure';
export const LoadingAuthType = 'scigateway:loading_auth';
export const LoadAuthProviderType = 'scigateway:loading_auth_provider';
export const LoadAuthUrlType = 'scigateway:loading_auth_url';
export const InvalidateTokenType = 'scigateway:api:invalidate_token';
export const ConfigureStringsType = 'scigateway:configure_strings';
export const LoginType = 'scigateway:login';
export const NotificationType = 'scigateway:api:notification';
export const RegisterRouteType = 'scigateway:api:register_route';
export const RequestPluginRerenderType = 'scigateway:api:plugin_rerender';
export const SignOutType = 'scigateway:signout';
export const ToggleDrawerType = 'scigateway:toggledrawer';
export const DismissNotificationType = 'scigateway:dismissnotification';
export const ConfigureFeatureSwitchesType = 'scigateway:feature_switches';
export const AddNotificationsType = 'scigateway:add_notification';
export const SiteLoadingType = 'scigateway:site_loading';
export const ConfigureAnalyticsType = 'scigateway:configure_analytics';
export const InitialiseAnalyticsType = 'scigateway:initialise_analytics';
export const ToggleHelpType = 'scigateway:toggle_help';
export const AddHelpTourStepsType = 'scigateway:add_help_tour_steps';

export interface NotificationPayload {
  message: string;
  severity: string;
}

export interface AddNotificationsPayload {
  notifications: NotificationPayload[];
}

export interface AppStrings {
  [id: string]: string;
}

export interface ConfigureStringsPayload {
  res: ApplicationStrings;
}

export interface ApplicationStrings {
  [section: string]: AppStrings;
}

export interface FeatureSwitchesPayload {
  switches: FeatureSwitches;
}

export interface FeatureSwitches {
  showContactButton: boolean;
}

export interface RegisterRoutePayload {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
}

export interface PluginConfig {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
}

export interface GroupedPlugins {
  [section: string]: PluginConfig[];
}

export interface LoginPayload {
  username: string;
  password: string;
}

export interface AccessPayload {
  authorisation: boolean;
  accessToken: string;
  error: null;
}

export interface AuthorisedPayload {
  token: string;
}

export interface DismissNotificationPayload {
  index: number;
}

export interface AuthProviderPayload {
  authProvider: string;
  authUrl: string | undefined;
}

export interface SiteLoadingPayload {
  loading: boolean;
}

export interface ConfigureAnalyticsPayload {
  id: string;
}

export interface AddHelpTourStepsPayload {
  steps: Step[];
}
