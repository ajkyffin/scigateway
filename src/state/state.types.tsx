import { ThunkAction } from 'redux-thunk';
import { AnyAction } from 'redux';
import { Step } from 'react-joyride';
import {
  ApplicationStrings,
  PluginConfig,
  FeatureSwitches,
} from './scigateway.types';
import { RouterState } from 'connected-react-router';

export interface Plugin {
  name: string;
  src: string;
  enable: boolean;
  location: 'main' | 'left' | 'right';
}

export interface ScigatewayNotification {
  message: string;
  severity: string;
}

export interface ScigatewayState {
  notifications: ScigatewayNotification[];
  plugins: PluginConfig[];
  drawerOpen: boolean;
  siteLoading: boolean;
  showHelp: boolean;
  helpSteps: Step[];
  authorisation: AuthState;
  res?: ApplicationStrings;
  features: FeatureSwitches;
  analytics?: AnalyticsState;
}

export interface StateType {
  scigateway: ScigatewayState;
  router: RouterState;
}

export interface ActionType<T> {
  type: string;
  payload: T;
}

export type ThunkResult<R> = ThunkAction<R, StateType, null, AnyAction>;

export interface User {
  username: string;
  avatarUrl: string;
}

// eslint-disable-next-line @typescript-eslint/interface-name-prefix
export interface ICATAuthenticator {
  mnemonic: string;
  keys: { name: string; hide?: boolean }[];
  friendly?: string;
  admin?: boolean;
}

export interface AuthProvider {
  isLoggedIn: () => boolean;
  logOut: () => void;
  logIn: (username: string, password: string) => Promise<void>;
  verifyLogIn: () => Promise<void>;
  refresh: () => Promise<void>;
  redirectUrl: string | null;
  user: User | null;
  mnemonic?: string;
  autoLogin?: () => Promise<void>;
}

export interface AuthState {
  failedToLogin: boolean;
  signedOutDueToTokenInvalidation: boolean;
  loading: boolean;
  provider: AuthProvider;
}

export interface AnalyticsState {
  id: string;
  initialised: boolean;
}
