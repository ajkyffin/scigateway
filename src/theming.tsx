import { createMuiTheme } from '@material-ui/core/styles';
import { ThemeOptions, Theme } from '@material-ui/core/styles/createMuiTheme';

export interface UKRIThemeOptions extends ThemeOptions {
  ukri: {
    purple: string;
    blue: string;
    orange: string;
    green: string;
    grey: string;
  };
}

export interface UKRITheme extends Theme {
  ukri: {
    purple: string;
    blue: string;
    orange: string;
    green: string;
    grey: string;
  };
}

export const buildTheme = (): Theme => {
  const options: UKRIThemeOptions = {
    typography: {
      useNextVariants: true,
    },
    palette: {
      primary: {
        main: '#1D4F91', // dark blue
      },
      secondary: {
        main: '#63666A', //grey
      },
    },
    ukri: {
      purple: '#8C4799',
      blue: '#1D4F91',
      orange: '#C34613',
      green: '#008275',
      grey: '#63666A',
    },
  };

  return createMuiTheme(options);
};
