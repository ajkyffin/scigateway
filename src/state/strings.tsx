import { AppStrings } from './scigateway.types';
import { StateType } from './state.types';

export function getAppStrings(
  state: StateType,
  section: string
): AppStrings | undefined {
  return state.scigateway.res ? state.scigateway.res[section] : undefined;
}

export const getString = (res: AppStrings | undefined, key: string): string =>
  (res && res[key]) || key;
