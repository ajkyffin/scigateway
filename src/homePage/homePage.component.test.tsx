import React from 'react';
import { createMount, createShallow } from '@material-ui/core/test-utils';
import HomePage from './homePage.component';
import { MuiThemeProvider } from '@material-ui/core';
import { buildTheme } from '../theming';
import { StateType } from '../state/state.types';
import configureStore from 'redux-mock-store';
import { initialState } from '../state/reducers/scigateway.reducer';

describe('Home page component', () => {
  let shallow;
  let mount;
  let mockStore;
  let state: StateType;

  beforeEach(() => {
    shallow = createShallow({});
    mount = createMount();

    mockStore = configureStore();
    state = JSON.parse(JSON.stringify({ scigateway: initialState }));
  });

  afterEach(() => {
    mount.cleanUp();
  });

  const theme = buildTheme();

  it('homepage renders correctly', () => {
    const wrapper = shallow(
      <MuiThemeProvider theme={theme}>
        <HomePage store={mockStore(state)} />
      </MuiThemeProvider>
    );
    expect(
      wrapper
        .dive()
        .dive()
        .dive()
    ).toMatchSnapshot();
  });
});
