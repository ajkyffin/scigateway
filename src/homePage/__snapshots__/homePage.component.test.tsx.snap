// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page component homepage renders correctly 1`] = `
<div>
  <div
    className="HomePage-bigImage-2"
  >
    <WithStyles(Typography)
      className="HomePage-title-1"
      variant="h2"
    >
      title
    </WithStyles(Typography)>
  </div>
  <div
    className="HomePage-howItWorks-3"
  >
    <WithStyles(Typography)
      className="HomePage-howItWorksTitle-4"
      variant="h4"
    >
      how-label
    </WithStyles(Typography)>
    <WithStyles(Grid)
      container={true}
      spacing={24}
    >
      <WithStyles(Grid)
        className="HomePage-howItWorksGridItem-5"
        item={true}
        md={4}
        sm={12}
      >
        <WithStyles(Typography)
          className="HomePage-howItWorksGridItemTitle-6"
          variant="h5"
        >
          explore-label
        </WithStyles(Typography)>
        <img
          className="HomePage-howItWorksGridItemImage-7"
          src="explore.jpg"
        />
        <WithStyles(Typography)
          className="HomePage-howItWorksGridItemCaption-8"
          variant="body1"
        >
          explore-description
        </WithStyles(Typography)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        className="HomePage-howItWorksGridItem-5"
        item={true}
        md={4}
        sm={12}
      >
        <WithStyles(Typography)
          className="HomePage-howItWorksGridItemTitle-6"
          variant="h5"
        >
          analyse-label
        </WithStyles(Typography)>
        <img
          className="HomePage-howItWorksGridItemImage-7"
          src="analyse.jpg"
        />
        <WithStyles(Typography)
          className="HomePage-howItWorksGridItemCaption-8"
          variant="body1"
        >
          analyse-description
        </WithStyles(Typography)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        className="HomePage-howItWorksGridItem-5"
        item={true}
        md={4}
        sm={12}
      >
        <WithStyles(Typography)
          className="HomePage-howItWorksGridItemTitle-6"
          variant="h5"
        >
          record-label
        </WithStyles(Typography)>
        <img
          className="HomePage-howItWorksGridItemImage-7"
          src="record.jpg"
        />
        <WithStyles(Typography)
          className="HomePage-howItWorksGridItemCaption-8"
          variant="body1"
        >
          record-description
        </WithStyles(Typography)>
      </WithStyles(Grid)>
    </WithStyles(Grid)>
    <WithStyles(Typography)
      className="HomePage-strapline-9"
      variant="h6"
    >
      banner-one
    </WithStyles(Typography)>
    <WithStyles(Typography)
      className="HomePage-purpose-10"
      variant="h5"
    >
      banner-two
    </WithStyles(Typography)>
  </div>
</div>
`;
